contracts/registries/GroupIPAssetRegistry.sol

```diff
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.23;
+
+import { EnumerableSet } from "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+import { ReentrancyGuardUpgradeable } from "@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+
+import { IIPAccount } from "../interfaces/IIPAccount.sol";
+import { IGroupIPAssetRegistry } from "../interfaces/registries/IGroupIPAssetRegistry.sol";
+import { IGroupingModule } from "../interfaces/modules/grouping/IGroupingModule.sol";
+import { ProtocolPausableUpgradeable } from "../pause/ProtocolPausableUpgradeable.sol";
+import { Errors } from "../lib/Errors.sol";
+import { IPAccountStorageOps } from "../lib/IPAccountStorageOps.sol";
+
+/// @title GroupIPAssetRegistry
+/// @notice Manages the registration and tracking of Group IPA, including the group members and reward pools.
+abstract contract GroupIPAssetRegistry is
+    IGroupIPAssetRegistry,
+    ProtocolPausableUpgradeable,
+    ReentrancyGuardUpgradeable
+{
+    using IPAccountStorageOps for IIPAccount;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+
+    IGroupingModule public immutable GROUPING_MODULE;
+
+    /// @dev Storage structure for the GroupIPAssetRegistry
+    /// @custom:storage-location erc7201:story-protocol.GroupIPAssetRegistry
+    struct GroupIPAssetRegistryStorage {
+        mapping(address groupIpId => EnumerableSet.AddressSet memberIpIds) groups;
+        mapping(address ipId => address rewardPool) rewardPools;
+        // whitelisted group reward pools
+        mapping(address rewardPool => bool isWhitelisted) whitelistedGroupRewardPools;
+    }
+
+    // keccak256(abi.encode(uint256(keccak256("story-protocol.GroupIPAssetRegistry")) - 1)) & ~bytes32(uint256(0xff));
+    bytes32 private constant GroupIPAssetRegistryStorageLocation =
+        0x34c89140582ad641fa679f955c67d1a82028bef0953ade7c28b8194cf080d600;
+
+    modifier onlyGroupingModule() {
+        if (msg.sender != address(GROUPING_MODULE)) {
+            revert Errors.GroupIPAssetRegistry__CallerIsNotGroupingModule(msg.sender);
+        }
+        _;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(address groupingModule) {
+        GROUPING_MODULE = IGroupingModule(groupingModule);
+        _disableInitializers();
+    }
+
+    /// @notice Registers a Group IPA
+    /// @param groupNft The address of the group IPA
+    /// @param groupNftId The id of the group IPA
+    /// @param rewardPool The address of the group reward pool
+    /// @return groupId The address of the newly registered Group IPA.
+    function registerGroup(
+        address groupNft,
+        uint256 groupNftId,
+        address rewardPool
+    ) external nonReentrant onlyGroupingModule whenNotPaused returns (address groupId) {
```

Comment:
should nonReentrant be here on on the entry function in grouping module?

Reply:
yes. move nonReentrant to entry point function.